infer/src/IR/Tenv.ml

```diff
@@ -81,51 +91,54 @@ let add_field tenv class_tn_name field =
 
 (** Reorder Hack's supers so that traits come first. *)
 let reorder_hack_supers tenv ~ignore_require_extends ({Struct.supers} as str) =
-  let supers = List.map supers ~f:(fun super -> (super, lookup tenv super)) in
+  let supers_with_lookup = List.map supers ~f:(fun super -> (super, lookup tenv super)) in
   let supers =
     if ignore_require_extends && Struct.is_hack_trait str then
-      List.filter supers ~f:(fun (_, str) ->
-          Option.exists str ~f:(fun str ->
-              Struct.is_hack_interface str || Struct.is_hack_trait str ) )
-    else supers
+      List.filter supers_with_lookup ~f:(fun (_, str_opt) ->
+          Option.exists str_opt ~f:(fun s ->
+              Struct.is_hack_interface s || Struct.is_hack_trait s ) )
+    else supers_with_lookup
   in
   let traits, others =
-    List.partition_tf supers ~f:(fun (_, str) -> Option.exists str ~f:Struct.is_hack_trait)
+    List.partition_tf supers ~f:(fun (_, str_opt) -> Option.exists str_opt ~f:Struct.is_hack_trait)
   in
   List.map traits ~f:fst @ List.map others ~f:fst
 
 
+(** Iterative fold over supers to avoid deep recursion.
+    The function f is applied to the current name and the optional struct. *)
 let fold_supers ?(ignore_require_extends = false) tenv name ~init ~f =
   let is_hack = Typ.Name.Hack.is_class name in
-  let rec aux worklist visited result =
-    match worklist with
-    | [] ->
-        (visited, result)
-    | name :: worklist when Typ.Name.Set.mem name visited ->
-        aux worklist visited result
-    | name :: worklist -> (
-        let visited = Typ.Name.Set.add name visited in
-        let struct_opt = lookup tenv name in
-        let result = f name struct_opt result in
-        match struct_opt with
-        | None ->
-            aux worklist visited result
-        | Some ({supers} as str) ->
-            let supers =
-              if is_hack then reorder_hack_supers tenv ~ignore_require_extends str else supers
-            in
-            let visited, result = aux supers visited result in
-            aux worklist visited result )
-  in
-  aux [name] Typ.Name.Set.empty init |> snd
+  let visited = ref Typ.Name.Set.empty in
```

Comment:
I prefer the terminal recursive function instead of a while loop. Matter of taste ;-)